contracts/exchange/ibetEscrow.sol

```diff
@@ -0,0 +1,422 @@
+            escrow.token,
+            escrow.sender,
+            escrow.recipient,
+            escrow.amount,
+            escrow.agent
+        );
+
+        return true;
+    }
+
+    /// @notice エスクロー完了
+    /// @param _escrowId エスクローID
+    function finishEscrow(uint256 _escrowId)
+        public
+        returns (bool)
+    {
+        // チェック：エスクローIDが直近ID以下であること
+        require(
+            _escrowId <= EscrowStorage(storageAddress).getLatestEscrowId(),
+            "The escrowId must be less than or equal to the latest escrow ID."
+        );
+
+        Escrow memory escrow;
+        (escrow.token, escrow.sender, escrow.recipient,
+            escrow.amount, escrow.agent, escrow.status) =
+                EscrowStorage(storageAddress).getEscrow(_escrowId);
+
+        // チェック：エスクローが取消済みではないこと
+        require(
+            escrow.status == false,
+            "Escrow must be valid."
+        );
+
+        // チェック：msg.senderがエスクローのagentであること
+        require(
+            escrow.agent == msg.sender,
+            "msg.sender must be the agent of the escrow."
+        );
+
+        // 更新：残高
+        setBalance(
+            msg.sender,
+            escrow.token,
+            balanceOf(escrow.recipient, escrow.token).add(escrow.amount)
+        );
+
+        // 更新：エスクロー中数量
+        setDeposit(
+            msg.sender,
+            escrow.token,
+            depositOf(escrow.sender, escrow.token).sub(escrow.amount)
+        );
+
+        // 更新：エスクロー情報
+        EscrowStorage(storageAddress).setEscrow(
+            _escrowId,
+            escrow.token,
+            escrow.sender,
+            escrow.recipient,
+            escrow.amount,
+            escrow.agent,
+            true
+        );
+
+        // イベント登録
+        emit EscrowFinished(
+            _escrowId,
+            escrow.token,
+            escrow.sender,
+            escrow.recipient,
+            escrow.amount,
+            escrow.agent
+        );
+
+        return true;
+    }
+
+    /// @notice 全ての残高を引き出しする
+    /// @dev エスクローで拘束されているものは引き出しされない
+    /// @param _token トークンアドレス
+    /// @return 処理結果
+    function withdraw(address _token)
+        public
+        returns (bool)
+    {
+        uint256 balance = balanceOf(msg.sender, _token);
+
+        require(
+            balance > 0,
+            "The balance must be greater than zero."
+        );
+
+        // 更新処理：トークン引き出し（送信）
+        IbetStandardTokenInterface(_token).transfer(msg.sender, balance);
+        setBalance(msg.sender, _token, 0);
+
+        // イベント登録
+        emit Withdrawn(_token, msg.sender);
+
+        return true;
+    }
+
+    /// @notice Deposit Handler：デポジット処理
+    /// @param _from アカウントアドレス：残高を保有するアドレス
+    /// @param _value デポジット数量
+    function tokenFallback(address _from, uint _value, bytes memory /*_data*/)
+        public
+        override
+    {
+        setBalance(
+            _from,
+            msg.sender,
+            balanceOf(_from, msg.sender).add(_value)
+        );
```

Comment:
[IMO]i suspect that Event emit is needed after tokenFallback.

Reply:
You're right. I will add a new event.